components/authz-service/server/v2/rules_property_test.go

```diff
@@ -18,17 +21,511 @@ import (
+
+			},
+			createProjectAndRuleGen,
+		))
+
+	properties.Property("reports no rules when all are staged and staged not included in request",
+		prop.ForAll(
+			func(reqs projectAndRuleReq) bool {
+				defer testDB.Flush(t)
+
+				reportRules(t, reqs.rules)
+
+				_, err := cl.CreateProject(ctx, &reqs.CreateProjectReq)
+				if err != nil {
+					return reportErrorAndYieldFalse(t, err)
+				}
+
+				reqs.rules, err = createRules(ctx, cl, reqs.rules)
+				if err != nil {
+					return reportErrorAndYieldFalse(t, err)
+				}
+
+				resp, err := cl.ListRules(ctx, &api.ListRulesReq{IncludeStaged: false})
+				if err != nil {
+					return reportErrorAndYieldFalse(t, err)
+				}
+
+				return len(resp.Rules) == 0
+
+			},
+			createProjectAndRuleGen,
+		))
+
+	properties.Property("reports rules when all are applied",
+		prop.ForAll(
+			func(reqs projectAndRuleReq) bool {
+				defer testDB.Flush(t)
+
+				reportRules(t, reqs.rules)
+
+				_, err := cl.CreateProject(ctx, &reqs.CreateProjectReq)
+				if err != nil {
+					return reportErrorAndYieldFalse(t, err)
+				}
+
+				reqs.rules, err = createRules(ctx, cl, reqs.rules)
+				if err != nil {
+					return reportErrorAndYieldFalse(t, err)
+				}
+
+				cl.ApplyRulesStart(ctx, &api.ApplyRulesStartReq{})
+
+				resp, err := cl.ListRules(ctx, &api.ListRulesReq{})
+				if err != nil {
+					return reportErrorAndYieldFalse(t, err)
+				}
+
+				return rulesMatch(reqs.rules, resp)
+
+			},
+			createProjectAndRuleGen,
+		))
+
+	properties.TestingRun(t)
+}
+
+func TestUpdateRuleProperties(t *testing.T) {
+	ctx := context.Background()
+	cl, testDB, _, _, seed := testhelpers.SetupProjectsAndRulesWithDB(t)
+	properties := getGopterParams(seed)
+	_, _, createProjectAndRuleGen := getGenerators()
+
+	properties.Property("updates a staged rule and applies update confirming transition from staged to applied",
+		prop.ForAll(
+			func(reqs projectAndRuleReq) bool {
+				defer testDB.Flush(t)
+
+				_, err := cl.CreateProject(ctx, &reqs.CreateProjectReq)
+				if err != nil {
+					return reportErrorAndYieldFalse(t, err)
+				}
+
+				_, err = cl.CreateRule(ctx, &reqs.rules[0])
+				if err != nil {
+					return reportErrorAndYieldFalse(t, err)
+				}
+
+				updateReq := api.UpdateRuleReq{
+					Id:         reqs.rules[0].Id,
+					ProjectId:  reqs.rules[0].ProjectId,
+					Name:       reqs.rules[0].Name + " updated",
+					Type:       reqs.rules[0].Type,
+					Conditions: reqs.rules[0].Conditions,
+				}
+				rStaged, err := cl.UpdateRule(ctx, &updateReq)
+				if err != nil {
+					return reportErrorAndYieldFalse(t, err)
+				}
+
+				cl.ApplyRulesStart(ctx, &api.ApplyRulesStartReq{})
+
+				rApplied, err := cl.GetRule(ctx, &api.GetRuleReq{Id: reqs.rules[0].Id})
+				if err != nil {
+					return reportErrorAndYieldFalse(t, err)
+				}
+
+				return rApplied.Rule.Status == "applied" &&
+					ruleMatches(updateReq, *rApplied.Rule) &&
+					rStaged.Rule.Status == "" && // TODO: this is wrong; should be "staged"
```

Comment:
I will be fixing UpdateRule to fill in "staged".

Reply:
Will that require a new migration to fix the update query? If so, I've got a small PR (coming soon) fixing up some rule queries, so maybe we could roll your fix into that migration

Reply:
No, we are not actually returning an object from the DB layer, so just did it in postgres.go.